### ISSUE TITLE
Performance: Start transfer of objects while paginating through list operation responses

### ISSUE THREAD

**Issue by nkemnitz** [rclone/rclone#5859]
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/long-delay-when-copying-millions-of-tiny-objects/27834/12

#### What is your current rclone version (output from `rclone version`)?
`v1.57.0-DEV`

#### What problem are you are trying to solve?
When transferring a large amount of objects from a single subdirectory to an empty destination, Rclone will first collect the entire list of objects in the source directory before starting any actual transfer. For 50k objects on S3, that is about 6 seconds delay, for 100 million objects, it becomes over 3 hours delay without a single object being transferred.

#### How do you think rclone should be changed to solve that?
S3 List requests are already limited to 1000 objects per response. While rclone keeps paginating through the source directory, the already collected files could get transferred (or checked for existence on the destination). This will keep source and especially destination busy right from the start and would help users to spot issues/bottlenecks with their transfer early on.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by rlizzo** [comment]
This is really bothering me as well. 

I'd be happy to contribute a PR if someone from the core team can direct me to any nonobvious parts of the code relevant to this and comment on if there are any specific pitfalls which might make this difficult to implement this given rclone's current architecture?

**Comment by olefrost** [comment]
@rlizzo 

Not an expert, but I can share how I would try to clarify the logic/limitations using VSCode:
I would make a small test folder with e.g. 20 files, then create a debug configuration with these parameters: --checkers=1 --transfers=1 --s3-list-chunk=5 and then start debugging with a breakpoint at this function:
https://github.com/rclone/rclone/blob/b783f09fc6961e82f0bfb9dff2126c24dae891c9/backend/s3/s3.go#L2165

and find the pagination loop from there, either by single stepping down the call hierarchy or by setting breakpoint at higher level functions visible in the call stack. I guess it will be difficult if the pagination loop is at a lower level, because you then would need to change the core engine (march) and its interface to all backends.

**Comment by olefrost** [comment]
@rlizzo Sorry, the above breakpoint isn't quite what I wanted to suggest, here are right places:
https://github.com/rclone/rclone/blob/b783f09fc6961e82f0bfb9dff2126c24dae891c9/backend/s3/s3.go#L2408-L2417
https://github.com/rclone/rclone/blob/b783f09fc6961e82f0bfb9dff2126c24dae891c9/backend/s3/s3.go#L2444

**Comment by ncw** [comment]
There are several things going on here...

Firstly the `List` interface

https://github.com/rclone/rclone/blob/b783f09fc6961e82f0bfb9dff2126c24dae891c9/backend/s3/s3.go#L2417

Only returns a chunk of entries when done. This needs redesigning into something which looks more like the `ListR` interface

https://github.com/rclone/rclone/blob/b783f09fc6961e82f0bfb9dff2126c24dae891c9/backend/s3/s3.go#L2444

Now I have an idea on how to do this and how to make the changes in the 40 odd backends by adding a new `ListC` (list chunked) interface.

However if you are using `--fast-list` and thus the `ListR` interface then the listings will start immediately.

However if you are doing a sync, rclone requires the complete source and destination listings so it can sort them and compare them to see what is missing.

You can turn this off with `--no-check-dest`

      --no-check-dest          Don't check the destination, copy regardless

But that will copy files regardless of whether they exist on the destination or not.

In summary, I know about this problem and it isn't easy to fix!

**Comment by jgupta** [comment]
In my case, [mc](https://github.com/minio/mc) mirror has far better performance.

This is for initial copy/mirror between one large bucket and other empty bucket.

rclone copied only a few files (~ 800kb, 3-4 files) in first 10 minutes.
mc started coping almost immediately, speed started with 8MiB/s and after 10 minutes it was coping at 19.48 MiB/s (speed is still increasing).

**Comment by ncw** [comment]
@jgupta that gives me an idea...

For initial copy/mirror rclone won't have a destination directory it needs to sync with, so if it uses the `ListC` interface then it can start transferring immediately.

So maybe the `ListC` interface idea is a workable idea for fixing this problem after all....

**Comment by blaineam** [comment]
Running into an issue with an insanely limited memory footprint requirement that would be nice to have some sort of ListC way to limit how many items are in memory at a given time with the goal of implementing some form of pagination so I can iterate through all files in an insanely memory optimized way. (Need to keep memory below 18MB, hovering at 15.7MB doing nothing but running the Go runtime) small file count folders are working great as soon as I navigate to a large folder file count the os kills it.

**Comment by ncw** [comment]
Note that the `ListC` interface now exists, except we called it `ListP`. It has been implemented for all the bucket based backends (eg S3, Azureblob, GCS etc) and a few of the others.

This is used to cut out to an on disk sorting method to match up the source and destination. This is controlled by the `--list-cutoff` flag so @blaineam if you want to use the minimum memory then set this small.

      --list-cutoff int   To save memory, sort directory listings on disk above this threshold (default 1000000)

For instant start of transfers if doing an s3 to s3 transfer the `--assume-listings-sorted` flag in this beta can be used

[v1.72.0-beta.9185.a67d7666e.fix-assume-listings-sorted](https://beta.rclone.org/branch/fix-assume-listings-sorted/v1.72.0-beta.9185.a67d7666e.fix-assume-listings-sorted/) on branch [fix-assume-listings-sorted](https://github.com/rclone/rclone/tree/fix-assume-listings-sorted) (uploaded in 15-30 mins)

Between the introduction of `ListP` and the `--assume-listings-sorted` flag I'm going to declare this issue fixed once the `--assume-listings-sorted` flag is merged.

**Comment by panthony** [comment]
@ncw Hello 👋🏻 

Do you see any risk in using  `--assume-listings-sorted` whether it's true or not when using `rclone lsf -R`?

When I'm relying on `--files-from-raw`, rclone does not stream the source (S3, GCS, ..) anymore like "gsutil" or "s3" would but  seems to build a state in memory that may explode its memory usage (talking > 10Gb).

**Comment by ncw** [comment]
@panthony 
> Do you see any risk in using `--assume-listings-sorted` whether it's true or not when using `rclone lsf -R`?

The flag will only help with `sync`, `copy` or `move`.

Rclone should already be streaming the results of `lsf -R`

> When I'm relying on `--files-from-raw`, rclone does not stream the source (S3, GCS, ..) anymore like "gsutil" or "s3" would but seems to build a state in memory that may explode its memory usage (talking > 10Gb).

You'd better describe what you are trying to do and I'll see if I can help!

**Comment by panthony** [comment]
Maybe I should update rclone first - it's been a while I did not do it - to ensures the issue still exists then :)

For context, the issue I had:

https://forum.rclone.org/t/high-memory-usage-of-s3-during-listing/35376/10

It was due to the logic of "UsesDirectoryFilters".

At one point I tried to remove the check against it here:

```go
// Can't use ListR if...
	if doListR == nil || // ...no ListR
		fi.HaveFilesFrom() || // ...using --files-from
		maxLevel >= 0 || // ...using bounded recursion
		len(fi.Opt.ExcludeFile) > 0 || // ...using --exclude-file
		fi.UsesDirectoryFilters() { // ...using any directory filters
		return listRwalk(ctx, f, path, includeAll, maxLevel, listType, fn)
	}
```

But I suffered occasional hanging sync that I never had time to dig.
